backend/internetarchive/internetarchive.go

```diff
@@ -0,0 +1,1086 @@
+	503, // Service Unavailable/Slow Down - "Reduce your request rate"
+}
+
+// NewFs constructs an Fs from the path
+func NewFs(ctx context.Context, name, root string, m configmap.Mapper) (fs.Fs, error) {
+	// Parse config into Options struct
+	opt := new(Options)
+	err := configstruct.Set(m, opt)
+	if err != nil {
+		return nil, err
+	}
+
+	// Parse the endpoints
+	ep, err := url.Parse(opt.Endpoint)
+	if err != nil {
+		return nil, err
+	}
+	fe, err := url.Parse(opt.FrontEndpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	root = strings.Trim(root, "/")
+
+	f := &Fs{
+		name: name,
+		opt:  *opt,
+		ctx:  ctx,
+	}
+	f.setRoot(root)
+	f.features = (&fs.Features{
+		BucketBased: true,
+	}).Fill(ctx, f)
+
+	f.srv = rest.NewClient(fshttp.NewClient(ctx))
+	f.srv.SetRoot(ep.String())
+
+	f.front = rest.NewClient(fshttp.NewClient(ctx))
+	f.front.SetRoot(fe.String())
+
+	if opt.AccessKeyID != "" && opt.SecretAccessKey != "" {
+		auth := fmt.Sprintf("LOW %s:%s", opt.AccessKeyID, opt.SecretAccessKey)
+		f.srv.SetHeader("Authorization", auth)
+		f.front.SetHeader("Authorization", auth)
+	}
+
+	f.pacer = fs.NewPacer(ctx, pacer.NewS3(pacer.MinSleep(10*time.Millisecond)))
+
+	// test if the root exists as a file
+	_, err = f.NewObject(ctx, "/")
+	if err == nil {
+		f.setRoot(betterPathDir(root))
+		return f, fs.ErrorIsFile
+	}
+	return f, nil
+}
+
+// setRoot changes the root of the Fs
+func (f *Fs) setRoot(root string) {
+	f.root = strings.Trim(root, "/")
+}
+
+// Remote returns the remote path
+func (o *Object) Remote() string {
+	return o.remote
+}
+
+// ModTime is the last modified time (read-only)
+func (o *Object) ModTime(ctx context.Context) time.Time {
+	return o.modTime
+}
+
+// Size is the file length
+func (o *Object) Size() int64 {
+	return o.size
+}
+
+// Fs returns the parent Fs
+func (o *Object) Fs() fs.Info {
+	return o.fs
+}
+
+// Hash returns the hash value presented by IA
+func (o *Object) Hash(ctx context.Context, ty hash.Type) (string, error) {
+	if ty == hash.MD5 {
+		return o.md5, nil
+	}
+	if ty == hash.SHA1 {
+		return o.sha1, nil
+	}
+	if ty == hash.CRC32 {
+		return o.crc32, nil
+	}
+	return "", hash.ErrUnsupported
+}
+
+// Storable returns if this object is storable
+func (o *Object) Storable() bool {
+	return true
+}
+
+// SetModTime is not supported
+func (o *Object) SetModTime(ctx context.Context, t time.Time) (err error) {
+	bucket, reqDir := o.split()
+	if bucket == "" {
+		return fs.ErrorCantSetModTime
+	}
+	if reqDir == "" {
+		return fs.ErrorCantSetModTime
+	}
+
+	// https://archive.org/services/docs/api/md-write.html
+	var patch = []interface{}{
+		// we should drop it first to clear all rclone-provided mtimes
+		struct {
+			Op   string `json:"op"`
+			Path string `json:"path"`
+		}{"remove", "/rclone-mtime"},
+		struct {
+			Op    string `json:"op"`
+			Path  string `json:"path"`
+			Value string `json:"value"`
+		}{"add", "/rclone-mtime", swift.TimeToFloatString(t)},
```

Comment:
Is there a metadata parameter which is already in use to store modification times? If there is, we should make an effort to use that to be compatible.

I think we should not use the `swift.TimeToFloatString` format - that got used in swift and s3 for compatibility with existing tools, but it isn't a great date format.

If we are making our own date format then I usually use `time.RFC3339Nano`  which is a standard time format, contains the timezone and is accurate to the nanosecond.

Reply:
I doubt we can write to `mtime` field, which is set by server's choice. And if we ever can write to it, it would break everything on IA.
Also using `time.RFC3339Nano` seems a good idea.